dlt/destinations/impl/sqlalchemy/db_api_client.py

```diff
@@ -0,0 +1,415 @@
+        self.sqla_transaction = sqla_transaction
+
+    def commit_transaction(self) -> None:
+        if self.sqla_transaction.is_active:
+            self.sqla_transaction.commit()
+
+    def rollback_transaction(self) -> None:
+        if self.sqla_transaction.is_active:
+            self.sqla_transaction.rollback()
+
+
+def raise_database_error(f: TFun) -> TFun:
+    @wraps(f)
+    def _wrap_gen(self: "SqlalchemyClient", *args: Any, **kwargs: Any) -> Any:
+        try:
+            return (yield from f(self, *args, **kwargs))
+        except Exception as e:
+            raise self._make_database_exception(e) from e
+
+    @wraps(f)
+    def _wrap(self: "SqlalchemyClient", *args: Any, **kwargs: Any) -> Any:
+        try:
+            return f(self, *args, **kwargs)
+        except Exception as e:
+            raise self._make_database_exception(e) from e
+
+    if inspect.isgeneratorfunction(f):
+        return _wrap_gen  # type: ignore[return-value]
+    return _wrap  # type: ignore[return-value]
+
+
+class SqlaDbApiCursor(DBApiCursorImpl):
+    def __init__(self, curr: sa.engine.CursorResult) -> None:
+        # Sqlalchemy CursorResult is *mostly* compatible with DB-API cursor
+        self.native_cursor = curr  # type: ignore[assignment]
+        curr.columns
+
+        self.fetchall = curr.fetchall  # type: ignore[assignment]
+        self.fetchone = curr.fetchone  # type: ignore[assignment]
+        self.fetchmany = curr.fetchmany  # type: ignore[assignment]
+
+    def _get_columns(self) -> List[str]:
+        return list(self.native_cursor.keys())  # type: ignore[attr-defined]
+
+    # @property
+    # def description(self) -> Any:
+    #     # Get the underlying driver's cursor description, this is mostly used in tests
+    #     return self.native_cursor.cursor.description  # type: ignore[attr-defined]
+
+    def execute(self, query: AnyStr, *args: Any, **kwargs: Any) -> None:
+        raise NotImplementedError("execute not implemented")
+
+
+class DbApiProps:
+    # Only needed for some tests
+    paramstyle = "named"
+
+
+class SqlalchemyClient(SqlClientBase[Connection]):
+    external_engine: bool = False
+    dialect: sa.engine.interfaces.Dialect
+    dialect_name: str
+    dbapi = DbApiProps  # type: ignore[assignment]
+    migrations: Optional[MigrationMaker] = None  # lazy init as needed
+
+    def __init__(
+        self,
+        dataset_name: str,
+        staging_dataset_name: str,
+        credentials: SqlalchemyCredentials,
+        capabilities: DestinationCapabilitiesContext,
+        engine_args: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        super().__init__(credentials.database, dataset_name, staging_dataset_name, capabilities)
+        self.credentials = credentials
+
+        if credentials.engine:
+            self.engine = credentials.engine
+            self.external_engine = True
+        else:
+            self.engine = sa.create_engine(
```

Comment:
could we create engine when first connection is created? sometimes sql client is created to just use a few internal functions and never opens the connection